### ISSUE TITLE
Callout sticks to edge of screen when panning quickly

### ISSUE THREAD

**Issue by rmnblm** [mapbox/mapbox-gl-native#7982]
**Platform:** iOS 10.2
**Mapbox SDK version:** 3.4.1

With the merged PR https://github.com/mapbox/mapbox-gl-native/pull/6676 into 3.4.0, the callout stays open, but I experience a weird bug, see the following gif:

![out](https://cloud.githubusercontent.com/assets/5942764/22745243/2ffa8130-ee20-11e6-8a6e-1bc64674879e.gif)

(BTW: Can I somehow disable that the callout should stay open?)

**Comment by adampitchie** [comment]
This doesn't appear to be an issue in ios-v3.5.0-beta.1. I will try reproducing this with 3.4.1.

**Comment by rmnblm** [comment]
Yes of course. I open the callout manually in code like this

``` swift
func showCallout(at location: CLLocationCoordinate2D, title: String) {
  let annotation = MGLPointAnnotation()
  annotation.title = title
  annotation.coordinate = location
  mapView.addAnnotation(annotation)
  mapView.selectAnnotation(annotation, animated: true)
}
```

**Comment by adampitchie** [comment]
With the code you provided, I'm still unable to reproduce your issue in the simulator as well as on device with the default annotation view. Perhaps the issue has to do with your custom annotation image implementation.

**Comment by rmnblm** [comment]
Hi Adam,

could you please give me some contact information (Telegram, WhatsApp, Slack) so that we can talk about this issue. I don't have a custom annotation image nor do I have left/right callout accessories (I have, but I commented them out and the error still occurs).

I appreciate your help, thanks!

**Comment by rmnblm** [comment]
For people struggling with the same problem: I temporarily fixed this issue with this little workaround. Even though it is called every time the region is changing, it doesn't affect performance.

``` swift
func mapViewRegionIsChanging(_ mapView: MGLMapView) {
  if mapView.visibleAnnotations == nil, let annotations = mapView.annotations {
    mapView.removeAnnotations(annotations)
  }
}
```

@adampitchie Could you contact me by email once you have time?

**Comment by adampitchie** [comment]
@rmnblm sure

**Comment by 1ec5** [comment]
We update the callout view (either repositioning or dismissing it) on the `mbgl::MapChangeRegionWillChange`, `mbgl::MapChangeRegionWillChangeAnimated`, `mbgl::MapChangeDidFinishRenderingFrame`, and `mbgl::MapChangeDidFinishRenderingFrameFullyRendered` events. After panning quickly, the annotation goes out of view during a drift animation, during which we get `mbgl::MapChangeRegionIsChanging` events followed by a final `mbgl::MapChangeRegionDidChangeAnimated` event. I’m not sure why we wouldn’t get getting `mbgl::MapChangeDidFinishRenderingFrame` in this case, so that’s something to look into.

I think we should consider modifying `-notifyMapChange:` to instead call `-updateCalloutView` on `mbgl::MapChangeRegionIsChanging`, `mbgl::MapChangeRegionDidChange`, and `mbgl::MapChangeRegionDidChangeAnimated`, as the macOS SDK implementation does. (The macOS SDK doesn’t quite have the same case, however, because drifting is driven by AppKit instead of an mbgl transition.)

> Can I somehow disable that the callout should stay open?

If you create a custom callout view, it is possible to set `dismissesAutomatically` to `YES` (#7646). However, there isn’t a way to control the dismissal behavior for standard callout views. If you consider this option to be important, please open a separate issue.

**Comment by 1ec5** [comment]
@frederoni points out that #8022 fixed this issue. 🎉

**Comment by boundsj** [comment]
The fix in #8022 can be tested using the latest iOS beta [currently 3.5.0 beta 2](https://github.com/mapbox/mapbox-gl-native/releases/tag/ios-v3.5.0-beta.2)

**Comment by rmnblm** [comment]
Awesome, thanks!
